cli/src/movement_util.rs

```diff
@@ -0,0 +1,170 @@
+        }
+    }
+    fn direction(&self) -> String {
+        match self {
+            Direction::Next => "forward".to_string(),
+            Direction::Prev => "back".to_string(),
+        }
+    }
+
+    fn next_node_type(&self) -> String {
+        match self {
+            Direction::Next => "descendant".to_string(),
+            Direction::Prev => "ancestor".to_string(),
+        }
+    }
+
+    fn get_target_revset(
+        &self,
+        working_commit_id: &CommitId,
+        edit: bool,
+        has_conflict: bool,
+        change_offset: u64,
+    ) -> Result<Rc<RevsetExpression>, CommandError> {
+        let wc_revset = RevsetExpression::commit(working_commit_id.clone());
+        // If we're editing, start at the working-copy commit. Otherwise, start from
+        // its direct parent(s).
+        let start_revset = if edit {
+            wc_revset.clone()
+        } else {
+            wc_revset.parents()
+        };
+
+        let target_revset = match self {
+            Direction::Next => if has_conflict {
+                start_revset
+                    .children()
+                    .descendants()
+                    .filtered(RevsetFilterPredicate::HasConflict)
+                    .roots()
+            } else {
+                start_revset.descendants_at(change_offset)
+            }
+            .minus(&wc_revset),
+
+            Direction::Prev => {
+                if has_conflict {
+                    // If people desire to move to the root conflict, replace the `heads()` below
+                    // with `roots(). But let's wait for feedback.
+                    start_revset
+                        .parents()
+                        .ancestors()
+                        .filtered(RevsetFilterPredicate::HasConflict)
+                        .heads()
+                } else {
+                    start_revset.ancestors_at(change_offset)
+                }
+            }
+        };
+
+        Ok(target_revset)
+    }
+}
+
+pub fn get_target_commit(
+    ui: &mut Ui,
+    workspace_command: &WorkspaceCommandHelper,
+    direction: Direction,
+    working_commit_id: &CommitId,
+    edit: bool,
+    has_conflict: bool,
+    change_offset: u64,
+) -> Result<Commit, CommandError> {
+    let target_revset =
+        direction.get_target_revset(working_commit_id, edit, has_conflict, change_offset)?;
+    let targets: Vec<Commit> = target_revset
+        .evaluate_programmatic(workspace_command.repo().as_ref())?
+        .iter()
+        .commits(workspace_command.repo().store())
+        .try_collect()?;
+
+    let target = match targets.as_slice() {
+        [target] => target,
+        [] => {
+            // We found no descendant.
+            return Err(user_error(format!(
+                "No {} found {} commit{} {}",
+                direction.next_node_type(),
+                change_offset,
+                if change_offset > 1 { "s" } else { "" },
+                direction.direction(),
+            )));
```

Comment:
This message is hard to read and hard to internationalize. I would prefer if you make this two separate messages depending on direction instead (i.e. something like `let message = if direction == Direction::Next { ... } else { ... };`).

(I don't like the `if change_offset > 1 { "s" } else { "" }` bit either but that was already there.)

Reply:
> This message is hard to read and hard to internationalize. I would prefer if you make this two separate messages depending on direction instead (i.e. something like `let message = if direction == Direction::Next { ... } else { ... };`).
> 

Makes sense. Done.

> (I don't like the `if change_offset > 1 { "s" } else { "" }` bit either but that was already there.)

Since I've touched this anyway, I've refactored this out also, so I spell out "commit" or "commits" completely, so it should be friendly to internalization.

Also, there's the age old parenthesized plural like: `No ancestor found 1 commit(s) back`. We could go with that as well?

Reply:
> so I spell out "commit" or "commits" completely, so it should be friendly to internalization.

Not necessarily, actually. I've heard Polish (maybe other languages too) has a different form for 1, 2, and 3 :) I've personally been going with just "commits" (i.e. including "1 commits") until we do i18n for real. I don't really care what we do until then.

Reply:
lol! This is the real world we're talking about. It's all exceptions and a few rules ;-)

`commit(s)` it is!